fix(6794): option for handling of utf8 filenames with python2 executable

gbataille opened the pull request:
I know there is a lot of spilled ink on the web on whether this construct
```python
reload(sys)
sys.setdefaultencoding('utf8')
```
should be used or not BUT

- in several hours of search, I could not make it work in any other way
- I have made the code optional (with the `--with-utf8` option) so that it's the user choice to activate it or not (--> no regression)
- My production tiling 1000+ images per month has been running without issues with this code for 6 months
- I have checked the consistency with the system `LC_CTYPE` env variable because both the code and the LC_CTYPE need to be set (AFAIK)

The only worry I have is windows. I can confirm it works on mac and linux (at least Ubuntu 14.04) but I don't have access to a windows machine. According to my research, `LC_CTYPE` does exist on Windows too, so we should be fine

rouault commented:
Hum not sure I like this patch very much.

What's the issue *without* the patch ? Can't you fix things with bla.decode('UTF-8') or bla.encode('UTF-8') ?

gbataille commented:
Issue is described in the ticket --> you can't handle files having utf8
chars in the name

And no, as mentioned, I'm not able to solve this any other way. I don't
like it too much either but I don't know ehat else to do. I spent a lot of
time 6 months ago and I ended up with puttin that without issue in
production

(The better fix would be to upgrade to python 3)
On Fri, 27 Jan 2017 at 18:07, Even Rouault <notifications@github.com> wrote:

> Hum not sure I like this patch very much.
>
> What's the issue *without* the patch ? Can't you fix things with
> bla.decode('UTF-8') or bla.encode('UTF-8') ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/190#issuecomment-275717854>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76fK0cskTdo09rdA6-8AXi3s-1Wj5ks5rWiQ_gaJpZM4Lv_Mf>
> .
>
-- 
Greg

gbataille commented:
Looking a bit more inside, the strings are properly handled as unicode, without needing a `decode`. but then when it tries to write them into a file, it just blows. I just found a new thing `codecs.open(file, 'w', 'utf-8')` but it does not change anything :(

rouault commented:
@gbataille I've proposed another approach in https://trac.osgeo.org/gdal/ticket/6794

gbataille commented:
damn, the files need to be opened in binary for the encode to apply... in
retrospect it's almost obvious...

Well thanks for that. I'll adapt the PR and the tests.
Do we want to leave that optional to avoid regression or do you think
nowadays it's fine to force UTF8 on everybody (actually my understanding is
that python 3 does that).
I would leave a command line parameter to avoid trouble.... let me know

On Sat, Jan 28, 2017 at 10:21 PM, Even Rouault <notifications@github.com>
wrote:

> @gbataille <https://github.com/gbataille> I've proposed another approach
> in https://trac.osgeo.org/gdal/ticket/6794
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/190#issuecomment-275875486>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76V8HhV34vpanTt5w0AG2FSdzYqkeks5rW7FVgaJpZM4Lv_Mf>
> .
>

rouault commented:
I don't think we need an extra switch.

gbataille commented:
cool.

soory to be a pain, but one more thing. It seems that this method will
actually fail (even on files not having utf8 characters) if it is run in a
terminal that has a non-utf8 LC_CTYPE env variable.
My understanding of those things is that UTF8 is a superset of the past
encodings like Latin1 and that therefore there is little reason for present
day setup to not be in UTF8. But I'm not sure how right that is and I have
been burned in the past on encoding issues.

Do you think we should "force" the utf-8 encoding no matter what or we
should do some checking on the LC_CTYPE setup?


Example:
with your patch, setting the term LC_CTYPE to fr_FR.latin1

> python gdal/swig/python/scripts/gdal2tiles.py
> autotest/pyscripts/data/test_bounds_close_to_tile_bounds_x.vrt
> /tmp/out_small -z 11

will return an error

On Sun, Jan 29, 2017 at 11:17 AM, Even Rouault <notifications@github.com>
wrote:

> I don't think we need an extra switch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/190#issuecomment-275905010>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76TcJNzMLibjK9Bv45Q_mk3iruqf5ks5rXGc-gaJpZM4Lv_Mf>
> .
>

rouault commented:
I've never seen such plays with encoding in other Python code, so I don't think this is a good idea to play with that in GDAL code. I think pretty much anyone on Unix uses UTF-8 based locales since years. There might be issues with Windows though but that's another story.

gbataille commented:
So I took your solution (thanks!!), added the missing places, added some tests.

As far as I can tell this works well on linux/mac with both python 2 and 3
On windows (I don't know the platform) I tried at a friends. It does not work, but it did not work before and the first problem is that windows terminal is not handling UTF8 well. So I'm not sure how to do anything more there.

With a bad LC_CTYPE (non UTF-8 compatible), it seems it will not crash, BUT the content of the googlemaps.html and co files are going to have some odd characters where the input file name should be.

I'll add a warning in case the script is run with a "bad" LC_CTYPE so users have a clue what's happending

gbataille commented:
Hey Even, the build seems to have failed for a reason not linked to my change. I don't seem to have the right to relaunch a build. Could you please do so?

rouault commented on gdal/swig/python/scripts/gdal2tiles.py:
```diff
@@ -651,8 +651,15 @@ def __init__(self, arguments ):
 
-        # Output the results
+        # LC_CTYPE check
+        if 'UTF-8' not in os.environ.get("LC_CTYPE", ""):
```
Could we emit the warning only if the filename is not ASCII ? because otherwise on Windows the warning will be always emitted since LC_CTYPE will not be defined there

gbataille replied:
hum. I thought I had found on the web that LC_CTYPE would be there, for other application like postgres used it... anyway, I don't really know. that's a good idea. I need to figure out how to do that properly. Let me think about it

gbataille replied:
ok I have it. I'll add a couple of tests and I'll submit an updated PR.

Cheers

rouault commented:
Sorry for the yet another breakage in MSVC builds. Just fixed it per https://ci.appveyor.com/project/OSGeo/gdal/build/1.0.1117

Besides that, there are Python 3 failures related to your changes : https://s3.amazonaws.com/archive.travis-ci.org/jobs/203139576/log.txt
And MacOSX isn't happy in gdal2tiles test either for other reasons : https://api.travis-ci.org/jobs/203139583/log.txt?deansi=true

gbataille commented:
yup. Ok, I'll continue on it

On Sun, Feb 19, 2017 at 2:24 PM, Even Rouault <notifications@github.com>
wrote:

> Sorry for the yet another breakage in MSVC builds. Just fixed it per
> https://ci.appveyor.com/project/OSGeo/gdal/build/1.0.1117
>
> Besides that, there are Python 3 failures related to your changes :
> https://s3.amazonaws.com/archive.travis-ci.org/jobs/203139576/log.txt
> And MacOSX isn't happy in gdal2tiles test either for other reasons :
> https://api.travis-ci.org/jobs/203139583/log.txt?deansi=true
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/190#issuecomment-280918946>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76VXWqwNGfT0eH8oieQzvz0KumJTpks5reEKVgaJpZM4Lv_Mf>
> .
>

gbataille commented:
python 2 vs 3 is sometimes a nightmare...
+ I had a regression on the way I get the LC_CTYPE in the tests. Thought I had sorted that out before... anyway. New submission in progress

gbataille commented:
I'm stupid. fixed my issue but did not rebase. I have a build in progress for nothing that I don't seem to be able to cancel. There is another one behind with the rebase :/

rouault commented:
Thanks.

Committed in r37425. I'm a bit nervous about the Chinese characters being in the source code repository, but apparently that works...
